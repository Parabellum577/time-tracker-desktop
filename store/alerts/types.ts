import { ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IError } from '@services/types'

export type AlertsAction = ActionType<typeof actions>

export interface IAlert {
  alertType: 'success' | 'error' | 'warning' | 'coldstart'
  alertMessage: string | IError
}

export interface IAlertsState {
  alertsArray: IAlert[]
}
