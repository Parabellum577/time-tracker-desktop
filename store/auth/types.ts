import { ActionType } from 'typesafe-actions'
import { DeepReadonly } from 'utility-types'

import * as actions from './actions'
import { IError } from '@types'

export type AuthAction = ActionType<typeof actions>

export type IAuthState = DeepReadonly<{
  readonly authorized: boolean
  readonly error?: IError | null
  readonly loading: boolean
  readonly isLoadingData: boolean
}>
