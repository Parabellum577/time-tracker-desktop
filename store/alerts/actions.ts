import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
