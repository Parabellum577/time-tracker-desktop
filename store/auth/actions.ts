import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
