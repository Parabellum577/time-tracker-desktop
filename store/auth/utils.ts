import persistStore from '../persistStore'
import uuid = require('uuid')
import { callRetriedApi } from '@services/call-api'
import api from '@services/api'

export interface IAuthData {
  Token: string
  Expires: number
  Refresh: string
  TracingID?: string
}

export async function updateRefreshToken(refresh?: string) {
  const refreshToken = refresh || (getAuthData() && getAuthData().Refresh)
  const authData: IAuthData = await callRetriedApi(
    api.user.RefreshToken,
    {
      Refresh: refreshToken,
    },
    [300, 300],
  )

  setAuthData(authData)
  return authData
}

const authKey = process.env.STAGE === 'local' ? 'AUTH_DATA_LOCAL' : 'AUTH_DATA'

export const setAuthData = (data: IAuthData | null) => {
  let TracingID: string = persistStore.get('TRACING_ID')

  if (!TracingID) {
    TracingID = uuid.v4()
    TracingID = TracingID.replace(TracingID.slice(9, 13), 'desk')
    persistStore.set('TRACING_ID', TracingID)
  }

  if (data) {
    data.TracingID = TracingID
  }
  persistStore.set(authKey, data)
}

export const getAuthData = (): IAuthData | null => {
  return persistStore.get(authKey)
}
