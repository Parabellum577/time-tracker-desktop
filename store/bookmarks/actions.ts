import { createAsyncAction, createAction } from 'typesafe-actions'
import { IError, IFullIssue } from '@types'

export const getBookmarks = createAsyncAction(
  'bookmarks/GET_BOOKMARKS_REQUEST',
  'bookmarks/GET_BOOKMARKS_SUCCESS',
  'bookmarks/GET_BOOKMARKS_FAILURE',
)<void, IFullIssue[], IError>()

export const handleBookmark = createAsyncAction(
  'bookmarks/HANDLE_BOOKMARK_REQUEST',
  'bookmarks/HANDLE_BOOKMARK_SUCCESS',
  'bookmarks/HANDLE_BOOKMARK_FAILURE',
)<IFullIssue, IFullIssue, IError>()

export const clearBookmarks = createAction('plannings/CLEAR_BOOKMARKS', action => () => action())
