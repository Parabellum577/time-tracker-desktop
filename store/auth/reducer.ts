import { getType } from 'typesafe-actions'

import * as actions from './actions'
import { AuthAction, IAuthState } from './types'

const initialState: IAuthState = {
  authorized: false,
  error: null,
  loading: false,
  isLoadingData: false,
}

export default (state: IAuthState = initialState, action: AuthAction) => {
  switch (action.type) {
    case getType(actions.loginUser.request):
      return { ...state, loading: true }
    case getType(actions.loginUser.success):
      return { ...state, authorized: true, error: null, loading: false }
    case getType(actions.loginUser.failure):
      return { ...state, authorized: false, error: action.payload, loading: false }
    case getType(actions.loadingDataAfterLogin):
      return { ...state, isLoadingData: action.payload }
    case getType(actions.signOut):
      return { ...state, authorized: false, error: null, loading: false }
    default:
      return state
  }
}
