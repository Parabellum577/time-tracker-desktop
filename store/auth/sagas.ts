import { SagaIterator } from 'redux-saga'
import { all, call, fork, put, take, race, delay, select, cancel } from 'redux-saga/effects'
import { ActionType, getType } from 'typesafe-actions'
import { push } from 'connected-react-router'

import * as alertActions from '@store/alerts/actions'
import * as userActions from '@store/currentUser/actions'
import * as planningsActions from '@store/plannings/actions'
import * as bookmarksActions from '@store/bookmarks/actions'
import * as productivityActions from '@store/productivity/actions'
import * as projectsActions from '@store/projects/actions'
import * as recentActions from '@store/recents/actions'
import * as timelineActions from '@store/timeline/actions'
import * as screenshotActions from '@store/screenshot/actions'
import * as trackerActions from '@store/trackers/actions'
import * as timeTrackerActions from '@store/timeTracker/actions'
import * as synchronizationActions from '@store/synchronization/actions'
import * as actions from './actions'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'
import { setAuthData, getAuthData, IAuthData, updateRefreshToken } from './utils'
import api from '@api'
import { IRootState } from '@store/rootReducer'
import { IPlanning } from '@services/types'
import time from '@services/time'
import { getActivePlanning, getSyncState } from '@services/helpers'
import { createWindow, getWindow } from '@electronMain/src/helpers/browserWindow'
import { sendEvent } from '@electronMain/src/analytics/facade'
import { syncSpentSlices } from '@store/timeTracker/sagas'

const getAuthorizedFromStore = (state: IRootState) => state.auth.authorized
const isOnlineFromStore = (state: IRootState) => state.internetConnection.isOnline

export function* authSaga() {
  while (true) {
    let authData = getAuthData()
    const [credentialsAction, otsAction] = yield race([
      take(actions.loginUser.request), // Login by credentials
      take(actions.otsLogin.request), // Login by OTS
      take(getType(actions.loginUserViaOAuth.request)), // Login via OAuth
      call(() => (authData ? Promise.resolve() : new Promise(() => {}))), // Login by saved Refresh Token
    ])
    sendEvent('user', 'LOGIN_REQUEST')
    if (credentialsAction) {
      yield call(setAuthDataByCreds, credentialsAction.payload)
    } else if (otsAction) {
      yield call(setAuthDataByOTS, otsAction.payload)
    }

    authData = getAuthData()
    if (!authData) {
      yield put(push('/login'))
      continue
    }
    yield call(refreshAuth, authData.Refresh)
    if (!authData) {
      yield put(push('/login'))
      continue
    }
    yield put(userActions.getCurrentUser.request())
    const { error } = yield race({
      success: take(userActions.getCurrentUser.success),
      error: take(userActions.getCurrentUser.failure),
    })

    if (error) {
      yield put(
        alertActions.showAlertMessage.request({
          alertMessage: 'Oops! Something went wrong. Please relogin.',
          alertType: 'error',
        }),
      )
      continue
    }

    const taskDataFetchAfterLogin = yield fork(function*() {
      yield put(actions.loginUser.success())
      yield put(push('/main'))
      yield call(takeDataAfterLoading)
    })
    sendEvent('user', 'LOGIN_SUCCESS')
    yield race([call(refreshAuthToken), take(actions.signOut)])

    console.log('Logout')
    if (taskDataFetchAfterLogin) {
      cancel(taskDataFetchAfterLogin)
    }
    yield call(clearDataAfterLogout)
    yield put(push('/login'))
  }
}

function* setAuthDataByCreds({ Login, Password }: ILoginUserPayload): SagaIterator {
  try {
    const authData: IAuthData = yield call(api.user.LoginUser, {
      Login,
      Password,
    })
    yield call(setAuthData, authData)
    return authData
  } catch (error) {
    yield call(setAuthData, null)
    yield put(alertActions.showAlertMessage.request({ alertType: 'error', alertMessage: error.message }))
    yield put(actions.loginUser.failure(error))
    return null
  }
}

function* setAuthDataByOTS({ OTS }: ILoginOTSPayload): SagaIterator {
  const authorized = yield select(getAuthorizedFromStore)
  if (authorized) {
    return
  }

  try {
    const authData: IAuthData = yield call(api.user.OTSLogin, { OTS })
    yield call(setAuthData, authData)
    return authData
  } catch (error) {
    yield call(setAuthData, null)
    yield put(alertActions.showAlertMessage.request({ alertType: 'error', alertMessage: error.message }))
    yield put(actions.loginUser.failure(error))
    return null
  }
}

export function* takeDataAfterLoading(): SagaIterator {
  yield put(actions.loadingDataAfterLogin(true))
  yield fork(api.project.UpdateProjectsCache)
  yield all([
    put(planningsActions.getOpenPlannings.request()),
    put(planningsActions.getColdPlannings.request()),
    put(planningsActions.getClosedPlannings.request()),
    put(trackerActions.getUserTrackers.request()),
    put(timelineActions.getTimeline.request()),
    put(projectsActions.getProjects.request()),
    race([
      take(getType(planningsActions.getOpenPlannings.failure)),
      take(getType(planningsActions.getOpenPlannings.success)),
    ]),
    race([take(getType(projectsActions.getProjects.failure)), take(getType(projectsActions.getProjects.success))]),
    race([
      take(getType(planningsActions.getColdPlannings.failure)),
      take(getType(planningsActions.getColdPlannings.success)),
    ]),
    race([
      take(getType(planningsActions.getColdPlannings.failure)),
      take(getType(planningsActions.getColdPlannings.success)),
    ]),
    race([
      take(getType(planningsActions.getClosedPlannings.failure)),
      take(getType(planningsActions.getClosedPlannings.success)),
    ]),
    race([
      take(getType(trackerActions.getUserTrackers.failure)),
      take(getType(trackerActions.getUserTrackers.success)),
    ]),
    race([take(getType(timelineActions.getTimeline.failure)), take(getType(timelineActions.getTimeline.success))]),
    race([take(getType(userActions.getUserSettings.failure)), take(getType(userActions.getUserSettings.success))]),
  ])
  yield put(actions.loadingDataAfterLogin(false))
}

export function* refreshAuthToken(): SagaIterator {
  let authData = getAuthData()
  while (true) {
    if (!authData) {
      return
    }

    const isOnline: boolean = yield select(isOnlineFromStore)
    if (isOnline) {
      try {
        authData = yield call(refreshAuth, authData.Refresh)
        sendEvent('user', 'REFRESH_TOKEN')
      } catch (error) {
        console.error(error)
      }
    }
    const refreshAfter = authData ? (authData.Expires - time.now()) * 800 : 60000
    yield delay(refreshAfter > 0 ? refreshAfter : 60000)
  }
}

export function* refreshAuth(refresh: string): SagaIterator {
  try {
    const authData = yield call(updateRefreshToken, refresh)
    return authData
  } catch (e) {
    console.log('refreshAuth -> error:', e)
    console.log('refreshAuth -> error:', e.code, typeof e.code)
    const fatalCodes = [1, 102, 107, 206, 209, -32000]
    // INVALID_TOKEN
    // REFRESH_TOKEN_EXPIRED
    // USER_NOT_FOUND
    // MAIL_CONFIRMATION_EXPIRED
    // INVALID_REFRESH
    // crypto/rsa: verification error
    if (fatalCodes.includes(e.code)) {
      const message =
        e.code === 206
          ? 'Your account has expired... Please confirm your email to continue.'
          : 'Oops! Something went wrong. Please relogin.'
      yield put(
        alertActions.showAlertMessage.request({
          alertMessage: message,
          alertType: 'error',
        }),
      )
      const mainWindow = getWindow() || createWindow()
      mainWindow.show()
      mainWindow.focus()
      yield call(setAuthData, null)
      return null
    } else {
      return getAuthData()
    }
  }
}

export function* clearDataAfterLogout(): SagaIterator {
  const activePlanning: IPlanning = yield call(getActivePlanning)
  const isMaster: IPlanning = yield call(getSyncState)

  if (activePlanning && activePlanning.ID && isMaster) {
    yield put(planningsActions.stopPlanning.request())
    yield race([
      take(getType(planningsActions.stopPlanning.success)),
      take(getType(planningsActions.stopPlanning.failure)),
    ])
  }
  try {
    yield call(syncSpentSlices)
  } catch (e) {
    console.error('clearDataAfterLogout -> error', e)
  }
  yield put(planningsActions.clearPlannings())
  yield put(bookmarksActions.clearBookmarks())
  yield put(productivityActions.clearProductivities())
  yield put(projectsActions.clearProjects())
  yield put(recentActions.clearRecent())
  yield put(screenshotActions.deleteScreenshotStorages())
  yield put(trackerActions.clearUserTrackers())
  yield put(timeTrackerActions.removeAllSpentSlices())
  yield put(synchronizationActions.changeMasterStatus(true))
  yield call(setAuthData, null)
}
