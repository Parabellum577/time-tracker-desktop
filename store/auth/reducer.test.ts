import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import { IAuthState } from './types'
import auth from './reducer'

describe('[Store] Auth Reducer', () => {
  it('[Success] [Snapshot] Should handle loginUser', () => {
    const loginUserAction: ActionType<typeof actions> = {
      type: getType(actions.loginUser.success),
      payload: null,
    }

    const currentState: IAuthState = {
      authorized: true,
      error: null,
      loading: false,
      isLoadingData: false,
    }

    expect(auth(currentState, loginUserAction)).toMatchSnapshot()
  })

  it('[Failure] Should handle showAlertMessage', () => {
    const loginUserAction: ActionType<typeof actions> = {
      type: getType(actions.loginUser.failure),
      payload: {
        code: 666,
        message: 'SOME_ERROR',
      },
    }

    const currentState: IAuthState = {
      authorized: false,
      error: loginUserAction.payload,
      loading: false,
      isLoadingData: false,
    }

    expect(auth(currentState, loginUserAction)).toMatchSnapshot()
  })

  it('[Success] [Snapshot] Should handle signOut', () => {
    const signOutAction: ActionType<typeof actions> = {
      type: getType(actions.signOut),
    }

    const currentState: IAuthState = {
      authorized: false,
      error: null,
      loading: false,
      isLoadingData: false,
    }

    expect(auth(currentState, signOutAction)).toMatchSnapshot()
  })
})
