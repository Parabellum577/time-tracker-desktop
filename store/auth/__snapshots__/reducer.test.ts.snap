// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Auth Reducer [Failure] Should handle showAlertMessage 1`] = `
Object {
  "authorized": false,
  "error": Object {
    "code": 666,
    "message": "SOME_ERROR",
  },
  "isLoadingData": false,
  "loading": false,
}
`;

exports[`[Store] Auth Reducer [Success] [Snapshot] Should handle loginUser 1`] = `
Object {
  "authorized": true,
  "error": null,
  "isLoadingData": false,
  "loading": false,
}
`;

exports[`[Store] Auth Reducer [Success] [Snapshot] Should handle signOut 1`] = `
Object {
  "authorized": false,
  "error": null,
  "isLoadingData": false,
  "loading": false,
}
`;
